pymc_extras/statespace/models/DFM.py

```diff
@@ -0,0 +1,828 @@
+            + (k_endog if error_order > 0 else 0)
+            + (self.k_exog_states if self._exog else 0)
+        )
+
+        # Initialize the PyMCStateSpace base class.
+        super().__init__(
+            k_endog=k_endog,
+            k_states=k_states,
+            k_posdef=k_posdef,
+            verbose=verbose,
+            measurement_error=measurement_error,
+        )
+
+    @property
+    def param_names(self):
+        names = [
+            "x0",
+            "P0",
+            "factor_loadings",
+            "factor_ar",
+            "error_ar",
+            "error_sigma",
+            "sigma_obs",
+        ]
+
+        # Handle cases where parameters should be excluded  or included based on model settings
+        if self.factor_order == 0:
+            names.remove("factor_ar")
+        if self.error_order == 0:
+            names.remove("error_ar")
+        if self.error_cov_type == "unstructured":
+            names.remove("error_sigma")
+            names.append("error_cov")
+        if not self.measurement_error:
+            names.remove("sigma_obs")
+        if self._exog:
+            names.append("beta")
+            if self.exog_innovations:
+                names.append("beta_sigma")
+
+        return names
+
+    @property
+    def param_info(self) -> dict[str, dict[str, Any]]:
+        info = {
+            "x0": {
+                "shape": (self.k_states,),
+                "constraints": None,
+            },
+            "P0": {
+                "shape": (self.k_states, self.k_states),
+                "constraints": "Positive Semi-definite",
+            },
+            "factor_loadings": {
+                "shape": (self.k_endog, self.k_factors),
+                "constraints": None,
+            },
+            "factor_ar": {
+                "shape": (self.k_factors, self.factor_order * self.k_factors),
+                "constraints": None,
+            },
+            "error_ar": {
+                "shape": (
+                    self.k_endog,
+                    self.error_order * self.k_endog if self.error_var else self.error_order,
+                ),
+                "constraints": None,
+            },
+            "error_sigma": {
+                "shape": (self.k_endog,) if self.error_cov_type == "diagonal" else (),
+                "constraints": "Positive",
+            },
+            "error_cov": {
+                "shape": (self.k_endog, self.k_endog),
+                "constraints": "Positive Semi-definite",
+            },
+            "sigma_obs": {
+                "shape": (self.k_endog,),
+                "constraints": "Positive",
+            },
+            "beta": {
+                "shape": (self.k_exog_states if self.k_exog is not None else 0,),
+                "constraints": None,
+            },
+            "beta_sigma": {
+                "shape": (self.k_exog_states if self.k_exog is not None else 0,),
+                "constraints": "Positive",
+            },
```

Comment:
These shapes should also change base on `shared_exog_state` I think

Reply:
also k_exog should never be `None`, you should check that `self.k_exog > 0`, or you use `_exog` flag